### PR TITLE
EdgeConnect config enhancements for k8s automation.

### PR DESCRIPTION
**Pull request by andriisoldatenko**
## Description

This PR adds an extra service account certificate and adds a `secrets` section to the config to EC config.

## How can this be tested?

`make go/test`

Deploy EC to your cluster:
```
apiVersion: dynatrace.com/v1alpha1
kind: EdgeConnect
metadata:
  name: e2e-andrii
  namespace: dynatrace
spec:
  apiServer: "<tenant>.dev.apps.example.com"
  replicas: 1
  oauth:
    clientSecret: edge-connect-client-secret-<tenant>
    endpoint: https://sso-dev.example.com/sso/oauth2/token
    resource: urn:dtenvironment:<tenannt>
```

and connect using debug pod and check yaml and EC logs.

edgeConnect.yaml:
```
k get secrets e2e-andrii-ec-yaml -o jsonpath="{.data['edgeConnect\.yaml']}" | base64 -d
name: e2e-andrii
api_endpoint_host: ******.dev.apps.******.com
oauth:
    endpoint: https://sso-dev.******.com/sso/oauth2/token
    client_id: ******
    client_secret: ******.******.******
    resource: urn:dtenvironment:******
root_certificate_paths:
    - /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
secrets:
    - name: K8S_SERVICE_ACCOUNT_TOKEN
      token: ******.******.******
      from_file: /var/run/secrets/kubernetes.io/serviceaccount/token
      restrict_hosts_to:
        - kubernetes.default.svc
```

### CODE REVIEW

File: pkg/controllers/edgeconnect/secret/secret.go [hunk 1]
```diff
@@ -80,3 +92,17 @@ func safeEdgeConnectCfg(cfg config.EdgeConnect) string {
 
 	return string(safe)
 }
+
+func addDefaultServiceAccount() (config.Secret, error) {
+	newToken, err := dttoken.New("dt0e01")
```
Comment (StefanHauth) [inline pkg/controllers/edgeconnect/secret/secret.go:97]: How is it made sure, that excactly this token is also used in the connection settings object?

File: pkg/controllers/edgeconnect/secret/secret.go [hunk 2]
```diff
@@ -80,3 +92,17 @@ func safeEdgeConnectCfg(cfg config.EdgeConnect) string {
 
 	return string(safe)
 }
+
+func addDefaultServiceAccount() (config.Secret, error) {
```
Comment (StefanHauth) [inline pkg/controllers/edgeconnect/secret/secret.go:96]: It's not necessarily the default service account. It's the token of whatever service account is configured.
```suggestion
func createServiceAccountTokenSecret() (config.Secret, error) {
```

**Review by StefanHauth — CHANGES_REQUESTED** [review]

File: pkg/controllers/edgeconnect/secret/secret.go (hunk 1 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/secret/secret.go:97]: yeah, it's dependency on another PR. will discuss with @albertogdd

File: pkg/controllers/edgeconnect/secret/secret.go (hunk 2 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/secret/secret.go]: fixed in 50f5f58f

File: pkg/controllers/edgeconnect/secret/secret.go [hunk 3]
```diff
@@ -80,3 +85,12 @@ func safeEdgeConnectCfg(cfg config.EdgeConnect) string {
 
 	return string(safe)
 }
+
+func createDefaultServiceAccount(token string) config.Secret {
```
Comment (StefanHauth) [inline pkg/controllers/edgeconnect/secret/secret.go:89]:
```suggestion
func createKubernetesApiSecret(token string) config.Secret {
```

File: pkg/controllers/edgeconnect/secret/secret.go [hunk 4]
```diff
@@ -46,6 +46,11 @@ func PrepareConfigFile(ctx context.Context, instance *edgeconnectv1alpha1.EdgeCo
 		cfg.RootCertificatePaths = append(cfg.RootCertificatePaths, consts.EdgeConnectMountPath+"/"+consts.EdgeConnectCustomCertificateName)
 	}
 
+	if instance.Spec.ServiceAccountName != "" {
+		cfg.RootCertificatePaths = append(cfg.RootCertificatePaths, consts.EdgeConnectServiceAccountCAPath)
+		cfg.Secrets = append(cfg.Secrets, createDefaultServiceAccount(token))
```
Comment (StefanHauth) [inline pkg/controllers/edgeconnect/secret/secret.go:51]: The secret shall only be added if automation is enabled.

**Review by StefanHauth — CHANGES_REQUESTED** [review]

File: pkg/controllers/edgeconnect/secret/secret.go [hunk 5]
```diff
@@ -46,6 +46,11 @@ func PrepareConfigFile(ctx context.Context, instance *edgeconnectv1alpha1.EdgeCo
 		cfg.RootCertificatePaths = append(cfg.RootCertificatePaths, consts.EdgeConnectMountPath+"/"+consts.EdgeConnectCustomCertificateName)
 	}
 
+	if instance.Spec.ServiceAccountName != "" {
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/secret/secret.go:49]: thanks @aorcholski! i'll remove it

File: pkg/controllers/edgeconnect/secret/secret.go (hunk 3 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/secret/secret.go:89]: fixed.

File: pkg/controllers/edgeconnect/secret/secret.go (hunk 4 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/secret/secret.go:51]: fixed in 3fa49162

File: pkg/controllers/edgeconnect/secret/secret.go [hunk 6]
```diff
@@ -46,6 +46,11 @@ func PrepareConfigFile(ctx context.Context, instance *edgeconnectv1alpha1.EdgeCo
 		cfg.RootCertificatePaths = append(cfg.RootCertificatePaths, consts.EdgeConnectMountPath+"/"+consts.EdgeConnectCustomCertificateName)
 	}
 
+	if instance.IsK8SAutomationEnabled() {
+		cfg.RootCertificatePaths = append(cfg.RootCertificatePaths, consts.EdgeConnectServiceAccountCAPath)
```
Comment (StefanHauth) [inline pkg/controllers/edgeconnect/secret/secret.go:50]: The certificate should always be added. Only the secret shall be only added if automation is enabled.

File: pkg/controllers/edgeconnect/secret/secret.go (hunk 3 above)
Comment (StefanHauth) [reply pkg/controllers/edgeconnect/secret/secret.go:89]: I don't see any change. Has something slipped the commit?

**Review by StefanHauth — CHANGES_REQUESTED** [review]

File: pkg/controllers/edgeconnect/secret/secret.go (hunk 3 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/secret/secret.go]: my bad., fixed in a4f78fbd

File: pkg/controllers/edgeconnect/secret/secret.go (hunk 6 above)
Comment (andriisoldatenko) [reply pkg/controllers/edgeconnect/secret/secret.go]: fixed in 70372af6

**Review by StefanHauth — APPROVED** [review]
LGTM

File: pkg/controllers/edgeconnect/config/config.go [hunk 7]
```diff
@@ -20,6 +20,9 @@ type EdgeConnect struct {
 
 	// Proxy related section.
 	Proxy Proxy `yaml:"proxy,omitempty"`
+
+	// Secrets related section.
+	Secrets []Secret `yaml:"secrets,omitempty"`
```
Comment (aorcholski) [inline pkg/controllers/edgeconnect/config/config.go:25]: is an array required?

**Review by aorcholski — APPROVED** [review]

**Pull request merged**